resource_providers/src/http.rs

```diff
@@ -0,0 +1,176 @@
+pub use self::bindgen::mycelia_alpha::http::types::{ClientResponse, Method};
+
+// this helps with syntax completion as it isn't entirely obvious what's happening under the hood
+// if you get really stuck you can use `cargo expand >> expanded.rs` to view the generated code
+mod bindgen {
+    use wasmtime::component::*;
+
+    bindgen!({
+      path: "../guest_crates/mycelia_http/wit",
+      world: "command",
+      async: true
+    });
+}
+
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+/// Errors which might occur when creating a new http client for a guest
+pub enum ClientMakeError {
+    #[error("wasm guest resource not found.")]
+    NotFound,
+    #[error("attempted to associate a resource to an existing id")]
+    BadResourceId,
+}
+
+#[derive(Error, Debug)]
+/// Errors which might occur when the host is attempting to make a request on behalf
+/// of a guest
+pub enum HttpClientError {
+    #[error("host client isn't ready. wait and try again.")]
+    NotReady,
+    #[error("unknown failure")]
+    Unknown,
+    #[error("client error")]
+    ClientError { cause: String },
+    #[error("guest produced a malformed request")]
+    BadRequest,
+    /// This should never happen. If we cannot find a client the guest is
+    /// either somehow intentionally requesting access to resources it should not access
+    /// Or, more likely we have a bug tracking the lifetime of http client resources
+    #[error("guest tried to use a client which does not exist")]
+    HostResourceNotFound,
+}
+
+/// Abstract service type defining an HttpClient
+/// which can be provided by specific concrete implementations
+/// for example we have a `HyperHostClient`
+pub type HostClient = BoxService<ClientRequest, ClientResult, HttpClientError>;
+
+/// Abstract service type for a thing which produces new HostClients
+/// for example see `providers::hyper::new_client_maker`
+pub type HostClientMaker = BoxService<(), HostClient, ClientMakeError>;
+
+/// Manages the association between guest wasm http clients and their host implementation instances.
+pub struct HostClientResource {
+    pub resource_id_provider: HostResourceIdProvider,
+    pub client_maker: BoxService<(), HostClient, ClientMakeError>,
+    pub clients: HashMap<u32, HostClient>,
+}
+
+impl HostClientResource {
+    pub fn new(
+        client_maker: HostClientMaker,
+        resource_id_provider: HostResourceIdProvider,
+    ) -> Self {
+        Self {
+            resource_id_provider,
+            client_maker,
+            clients: Default::default(),
+        }
+    }
+}
+
+#[async_trait]
+impl HostClientInterface for HostClientResource {
+    /// Creates a new HttpClient Resource and returns the resource to the guest.
+    ///
+    /// We return a resource id the guest will use to request access to the resource
+    /// internally, we create a mapping of <ID, Client> so we can lookup the correct
+    /// client to make the request.
```

Comment:
This ID appears to live in the session, is that correct? Should it be backed up for data integrity reasons?

Reply:
These resource IDs are ephemeral